api/v1alpha1/api_key_auth_types.go

```diff
@@ -0,0 +1,69 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+const APIKeysSecretKey = "credentials"
+
+// APIKeyAuth defines the configuration for the API Key Authentication.
+type APIKeyAuth struct {
+	// Credentials is the Kubernetes secret which contains the API keys.
+	// This is an Opaque secret.
+	// Each API key is stored in the key representing the client id,
+	// which can be used in AllowedClients to authorize the client in a simple way.
+	// If this is attached to a HTTPRoute while the other SecurityPolicies with APIKeyAuth is attached to a Gateway,
+	// this credentials takes precedence.
+	// This field is optional, but it doesn't make sense that any attached APIKeyAuth for the route/gateway doesn't have credentials at all.
+	//
+	// +optional
+	Credentials gwapiv1.SecretObjectReference `json:"credentials,omitempty"`
+
+	// KeySources is where to fetch the key from the coming request.
+	// The value from the first source that has a key will be used.
+	// If this is attached to a HTTPRoute while the other SecurityPolicies with APIKeyAuth is attached to a Gateway,
+	// this key sources takes precedence.
+	// This field is optional, but it doesn't make sense that any attached APIKeyAuth for the route/gateway doesn't have key sources at all.
+	//
+	// +optional
+	KeySources []*KeySource `json:"keySources,omitempty"`
```

Comment:
The `KeySources`  should not be optional.

The SecurityPolicy at HTTPRoute should override the SecurityPolicy at the Gateway as a whole.  The SecurityPolicy at the HTTPRoute should not use  the `KeySources` from the parent Gateway.